patches/v8/0019-Implement-cross-request-context-promise-resolve-hand.patch

```diff
@@ -0,0 +1,339 @@
++ * reactions to the resolved promise to be enqueued. The idea is that the
++ * embedder sets this callback in the case it needs to defer the actual
++ * scheduling of the reactions to the given promise to a later time.
++ * Importantly, when this callback is invoked, the state of the promise
++ * should have already been updated. We're simply possibly deferring the
++ * enqueue of the reactions to the promise.
++ */
++using PromiseCrossContextResolveCallback = void (*)(
++    v8::Isolate* isolate,
++    Local<Value> tag,
++    Local<Data> reactions,
++    Local<Value> argument,
++    std::function<void(v8::Isolate* isolate,
++                       Local<Data> reactions,
++                       Local<Value> argument)> callback);
++
+ }  // namespace v8
+ 
+ #endif  // INCLUDE_V8_ISOLATE_CALLBACKS_H_
+diff --git a/include/v8-isolate.h b/include/v8-isolate.h
+index 8112231319b..9fe57f7bde4 100644
+--- a/include/v8-isolate.h
++++ b/include/v8-isolate.h
+@@ -1732,6 +1732,8 @@ class V8_EXPORT Isolate {
+ 
+   class PromiseContextScope;
+   void SetPromiseCrossContextCallback(PromiseCrossContextCallback callback);
++  void SetPromiseCrossContextResolveCallback(
++      PromiseCrossContextResolveCallback callback);
+ 
+   Isolate() = delete;
+   ~Isolate() = delete;
+diff --git a/src/api/api.cc b/src/api/api.cc
+index d3fc6d77ea3..3255fd77e25 100644
+--- a/src/api/api.cc
++++ b/src/api/api.cc
+@@ -12246,6 +12246,11 @@ void Isolate::SetPromiseCrossContextCallback(PromiseCrossContextCallback callbac
+   isolate->set_promise_cross_context_callback(callback);
+ }
+ 
++void Isolate::SetPromiseCrossContextResolveCallback(PromiseCrossContextResolveCallback callback) {
++  i::Isolate* isolate = reinterpret_cast<i::Isolate*>(this);
++  isolate->set_promise_cross_context_resolve_callback(callback);
++}
++
+ Isolate::PromiseContextScope::PromiseContextScope(Isolate* isolate, v8::Local<v8::Object> tag)
+     : isolate_(reinterpret_cast<i::Isolate*>(isolate)) {
+   DCHECK(!isolate_->has_promise_context_tag());
+diff --git a/src/builtins/promise-abstract-operations.tq b/src/builtins/promise-abstract-operations.tq
+index fdec3d268e4..456d051d3c0 100644
+--- a/src/builtins/promise-abstract-operations.tq
++++ b/src/builtins/promise-abstract-operations.tq
+@@ -23,6 +23,9 @@ extern transitioning runtime PromiseRejectEventFromStack(
+ 
+ extern transitioning runtime PromiseContextCheck(
+     implicit context: Context)(JSPromise): JSPromise;
++
++extern transitioning runtime PromiseResolveContextCheck(
++    implicit context: Context)(JSPromise): JSAny;
+ }
+ 
+ // https://tc39.es/ecma262/#sec-promise-abstract-operations
+@@ -239,7 +242,8 @@ transitioning builtin RejectPromise(
+   // the runtime handle this operation, which greatly reduces
+   // the complexity here and also avoids a couple of back and
+   // forth between JavaScript and C++ land.
+-  if (IsIsolatePromiseHookEnabledOrDebugIsActiveOrHasAsyncEventDelegate(
```

Comment:
There are generally two paths for a promise to be resolved/rejected... one via runtime (logic in objects.cc) and one via these torque built-ins. There is existing logic on the torque side to always defer to the runtime logic when either promise hooks are being used or the inspector is enabled. In this patch we add a third case to this check to defer to the runtime logic if the promise tag does not match the current isolate promise tag.